### PR TITLE
modules: pin dependencies and fix updategenerated

### LINKED ISSUE #996: make updategenerated is broken after Go module changes
**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]

With a clean clone (no `vendor` folder`), `make updategenerated` fails because of

```
Updating CRD generated code...
hack/update-generated-crd-code.sh: line 22: hack/../vendor/k8s.io/code-generator/generate-groups.sh: No such file or directory
Makefile:95: recipe for target 'updategenerated' failed
make: *** [updategenerated] Error 127
```

**What did you expect to happen:**

`make updategenerated` passes without error

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Contour version:
- Kubernetes version: (use `kubectl version`):
- Kubernetes installer & version:
- Cloud provider or hardware configuration:
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by glerchundi**
Closes #996

Signed-off-by: Gorka Lerchundi Osa <glertxundi@gmail.com>

### CODE REVIEW

**Comment by davecheney** [comment]
@glerchundi thank you for your PR. 

While I am thankful for your PR, I prefer it to #997, @unicell already was working on this change and I don't want to encourage duelling PRs, that is not productive.

Please let me resolve #997 with @unicell and I'll review this changer afterwards.

**Comment by glerchundi** [comment]
Ups, i didn't noticed that there was a PR already sent. 

> @unicell already was working on this change and I don't want to encourage duelling PRs, that is not productive.

I agree, feel free to do whatever you want with this PR.

File: hack/update-generated-crd-code.sh [hunk 1]
```diff
@@ -15,13 +15,14 @@
...
 
-HACK_DIR=$(dirname "${BASH_SOURCE}")
-REPO_ROOT=${HACK_DIR}/..
-
-${REPO_ROOT}/vendor/k8s.io/code-generator/generate-groups.sh \
+VERSION=$(go list -m all | grep k8s.io/code-generator | rev | cut -d"-" -f1 | cut -d" " -f1 | rev)
+TMP_DIR=$(mktemp -d)
+git clone https://github.com/kubernetes/code-generator.git ${TMP_DIR}
+(cd ${TMP_DIR} && git reset --hard ${VERSION})
+${TMP_DIR}/generate-groups.sh \
```
Comment (unicell) [inline hack/update-generated-crd-code.sh]: Please note, if `contour` is checked out outside of $GOPATH, `code-generator` would still generate the CRD code into $GOPATH/src/github.com/heptio/contour/apis anyway due to https://github.com/kubernetes/code-generator/issues/57

Comment (glerchundi) [reply hack/update-generated-crd-code.sh]: Yep, but i think this is assumed as per the [contributing guidelines](https://github.com/heptio/contour/blob/master/CONTRIBUTING.md).

Comment (unicell) [reply hack/update-generated-crd-code.sh]: Cool, then we should be good

**Comment by stevesloka** [comment]
I'm trying to make some changes to the CRD and regen a new client but running into issues. 

I took the code from this PR and added to my project. When I run a `make updategenerated` I get the following output:

```
Generating deepcopy funcs
Generating clientset for contour:v1beta1 at github.com/heptio/contour/apis/generated/clientset
**EvalSymlinks: too many links**
Generating listers for contour:v1beta1 at github.com/heptio/contour/apis/generated/listers
Generating informers for contour:v1beta1 at github.com/heptio/contour/apis/generated/informers
```

Everything seems ok, but did notice an issue on client-go where a similar problem (i.e. too many links) was found and it was a version mismatch: https://github.com/kubernetes/code-generator/issues/37

**Comment by glerchundi** [comment]
@davecheney fixed by using `go mod init` (good catch!) and tested that it works in 1.12 also. I tried to avoid the use of `GO111MODULE` in order to be this a little bit more future-proof.

**Comment by glerchundi** [comment]
@stevesloka i cannot reproduce, can you look at that for me?

**Comment by stevesloka** [comment]
I don't think the $VERSION variable returns the right version, for me, it returns `generator`. We need to pin the generator to the same version as apimachinery / api & client-go. I set VERSION to `kubernetes-1.14.1` and it worked finally in addition to setting versions in go.mod as well for the previously mentioned

**Comment by unicell** [comment]
Since @stevesloka is mentioning the $VERSION variable, not sure if it is a concern, the code-generator dependency we're adding to the `go.mod` will be removed by `go mod tidy` as code-generator not really being referenced from the go code.

**Comment by glerchundi** [comment]
@stevesloka i tried to use the same version that appears on the go.mod. Go
modules support semver tags or directly the sha (prefixed by a
v0.0.0-somekindoftimestamp-sha). That snippet tries to extract the semver
or the sha. I cannot remember which version the sha I referenced on the go
modules points to but, if i’m not wrong, it was the same version used by
the other k8s dependencies, 1.12.6?

That being said, i dont know if there is a real problem in pointing the
generator to a newer version than the one used by the other k8s deps. as it
seems to fix, at least on your environment.

**Comment by glerchundi** [comment]
@unicell go mod tidy is not going to purge those deps as these appear in
the tools.go file this PR adds.

**Comment by stevesloka** [comment]
Testing this again I think this is working. I wiped my repo and started again and got my CRD's working. Thanks!

**Comment by glerchundi** [comment]
You're welcome, thank you all for making this project as strong and reliable as possible with these kind of reviews!

**Pull request merged**